Rewrite Rules API docs for call-by-name. (#22699)
pantsbuild/pants: docs/docs/writing-plugins/the-rules-api/concepts.mdx

    return str(i)
```

> Rules are typically module-level functions. In some cases you can define rules in nested scopes,  such as inside a class or function body. But this is useful only in specific, special cases in the Pants codebase, and you are unlikely to need to use this in practice.

Reviewer: ```suggestion
Rules are typically module-level functions. In some cases you can define rules in nested scopes, such as inside a class or function body. But this is useful only in specific, special cases in the Pants codebase, and you are unlikely to need to use this in practice.
```

> Although any immutable Python type, including builtin types like `int`, can be a parameter or a return type of a rule, in almost all cases rules will deal with values of custom Python classes. These are are typically implemented as [frozen dataclasses](https://docs.python.org/3/library/dataclasses.html).

Reviewer: > These are are typically implemented as [frozen dataclasses]

Any value in adding why they're this way? Why frozen, why a dataclass, etc? I've always just cargo-culted this, and even in some of the base classes - we say "Subclasses must set `@dataclass(frozen=True)` for their declared fields to be recognized.". Never really gave it much thought, but I miss adding them quite often.

Reviewer: This might also be addressed below tangentially? 
> The engine invokes rules concurrently where possible, to make use of all available local and remote cores. This is why rule params and return values must be immutable.

Reviewer: Also, mentioned before in the Dataclasses section. Another case of maybe foreshadow the why - as you have to read pretty far down to see (or even just a link anchor?)

Generally, a rule corresponds to a step in your build process. For example, when implementing a rule to run [shellcheck](https://www.shellcheck.net/) on a set of shell scripts, you could have a rule that maps `(Target, Shellcheck) -> LintResult`:

```python
@rule
async def run_shellcheck(target: Target, shellcheck: Shellcheck) -> LintResult:
    # Your logic.

>     return LintResult(stdout=..., stderr=.., exit_code=...)

Reviewer: ```suggestion
    return LintResult(stdout=..., stderr=..., exit_code=...)
```

- The engine applies memoization, so that if a rule has already run with the given params, the engine will supply the output value from the in-memory cache, instead of executing the rule. This is why rules must be [pure](https://en.wikipedia.org/wiki/Pure_function) and why rule params and return values must be hashable.

This requirement of rule purity is worth emphasizing: a rule must yield the same output for a given set of inputs, and a rule must not directly or indirectly rely on side-effecting code like `print()`, `subprocess.run()`, or `requests`. The Rules API provides alternatives that are understood by the Pants engine and which work properly with its caching and concurrency mechanisms.

## Invoking other rules in a rule body

One obvious way for a rule to depend on values of given types is to declare input parameters of those types. However it is very common to request extra values in the rule body by explicitly calling other rules. This is useful when you want programmatic control over the inputs to those other rules, or when you want to invoke other rules conditionally.

> To call a rule explicitly, you `await` it, and pass explicit and/or implicit params to it. The following snippet shows this in action (note that Pants ships with real shellcheck support that is more complicated, this example is simplified for clarity):

Reviewer: I find this example confusing specifically because of the double execute_process and it not being clear that the first one actually "does the thing", while the second basically just unwraps the result.

Maybe change:
> The following snippet shows this in action

to something more like:
> The following contrived example walks through the process of explicitly calling a sequence of rules

Or something like that?

Reviewer: Okay, in reading further down the page - I think we should allude to this being re-visited further down.

```python
from pants.engine.rules import rule
from pants.engine.intrinsics import execute_process
from pants.engine.process import fallible_to_exec_result_or_raise

@rule
async def run_shellcheck(target: Target, shellcheck: Shellcheck) -> LintResult:
    ...
    process_request = Process(
        ["/bin/echo", str(target.address)],
        description=f"Echo {target.address}",
    )

>     fallible_process_result = await execute_process(

Reviewer: ```suggestion
    fallible_process_result: FallibleProcessResult = await execute_process(
```

        process_request, **implicitly()
    )

>     process_result = await fallible_to_exec_result_or_raise(

Reviewer: Added a type to hopefully make it clear why there are two calls

        fallible_process_result, **implicitly()
    )
    return LintResult(
        stdout=process_result.stdout, stderr=process_result.stderr, exit_code=0
    )
```

The Pants engine will run your rule as straight-line Python code until it encounters the `await`, which will yield execution back to the engine. The engine will then see if it has a memoized result for the requested rule invocation. If not, it will execute the rule to obtain such a value.  Once the engine gives back the resulting output value, control will be returned back to your Python code,

> until the next `await`.

Reviewer: Should be merged with previous line?

- From the input parameters of the calling rule.
- Computed from other params by (transitively) applying suitable rules. You can think of this as a form of dependency injection via type: Pants knows the type of the implicit parameter, and can traverse a path through rule execution to go from an initial set of values, known from context, to the needed value.

Since explicit params must be provided positionally, they must be the first arguments to the rule. This means that when you write a rules, you should put the parameters expected to be passed explicitly before the parameters expected to be provided implicitly.

### Extra context for implicit parameters

As mentioned above, Pants can compute values for implicit parameters by transitively applying rules. In many cases the initial parameters for *those* rules are known from external context. But in some cases we need to provide extra context from the calling rule. To do so, we pass the contextual parameters as arguments to `**implicitly()`:

```python
from pants.engine.process import fallible_to_exec_result_or_raise
from pants.engine.rules import implicitly, rule

@rule
async def my_rule() -> MyResult:

>     process_result = await fallible_to_exec_result_or_raise(

Reviewer: ```suggestion
    process_result = await execute_process_or_raise(
```

        **implicitly(
            Process(
                ["/bin/echo", str(target.address)],
                description=f"Echo {target.address}",
            )
        )
    )
    ...
```

> In this example the `fallible_to_exec_result_or_raise()` rule takes a `FallibleProcessResult` and returns a `ProcessResult` by first checking the `FallibleProcessResult` for success and raising an exception if it failed. We saw this earlier, in the simplified shellcheck example.

Reviewer: ```suggestion
In this example the `execute_process_or_raise()` rule takes a `FallibleProcessResult` and returns a `ProcessResult` by first checking the `FallibleProcessResult` for success and raising an exception if it failed. We saw this earlier, in the simplified shellcheck example.
```

> But instead of explicitly passing a `FallibleProcessResult` as we did earlier, we now pass a `Process` as implicit context. The Pants engine then looks at all the rules it knows about to figure out how to compute a `FallibleProcessResult` from a `Process`. The `execute_process()` we encountered earlier fits the bill, and so the engine calls it on our `Process` and passes its return value into `fallible_to_exec_result_or_raise()`. Whereas earlier we called both rules explicitly, here we get the exact same behavior with just one call.

Reviewer: ```suggestion
But instead of explicitly passing a `FallibleProcessResult` as we did earlier, we now pass a `Process` as implicit context. The Pants engine then looks at all the rules it knows about to figure out how to compute a `FallibleProcessResult` from a `Process`. The `execute_process()` we encountered earlier fits the bill, and so the engine calls it on our `Process` and passes its return value into `execute_process_or_raise()`. Whereas earlier we called both rules explicitly, here we get the exact same behavior with just one call.
```

> In fact, since raising an exception on process failure is frequently what you want, we have an alias, `execute_process_or_raise`, to make the code more readable when using this common shorthand idiom.

Reviewer: I don't think this is necessary info, just ditch the old, verbose version - and we can clean up the Pants code to use `execute_process_or_raise` exclusively.

```python
from pants.engine.process import execute_process_or_raise
from pants.engine.rules import implicitly, rule

@rule
async def my_rule() -> MyResult:
    process = Process(...)
    ...
    process_result = await execute_process_or_raise(
        **implicitly({
            process: Process,
            ProductDescription("Running echo"): ProductDescription,
        })
    )
    ...
```

As you can see above, this also allows you to pass multiple contextual params to `**implicitly()`.

## Rule concurrency

The engine pauses execution on each `await` in your rule until the result is returned. This means that if you have two consecutive `await`s, the engine will evaluate them sequentially.

If your rules can be executed concurrently (because nether depends on the result of the other) then you can use `concurrently(...)` to instead get multiple results in a single `await`:

```python
from pants.engine.rules import rule

> from pants.engine.internals.selectors import concurrently

Reviewer: We have a split of this import, and `pants.engine.rules` as a re-export.

## Rule concurrency

The engine pauses execution on each `await` in your rule until the result is returned. This means that if you have two consecutive `await`s, the engine will evaluate them sequentially.

If your rules can be executed concurrently (because nether depends on the result of the other) then you can use `concurrently(...)` to instead get multiple results in a single `await`:

```python
from pants.engine.rules import concurrently, rule

@rule
async def lint_single_target(target: Target) -> LintResult:
    ...

@rule
async def lint_all(targets: Targets) -> LintResults:
    single_results = await concurrently(
        lint_single_target(target, **implicitly()) for target in targets
    )
    ...
```

The result of `concurrently` is a tuple with each individual result, in the same order as the requests. You should hardly ever call `await` in a loop - use `await concurrently` instead.

`concurrently` can either take an iterable of rule calls, as above, or take multiple individual rule calls. For example:

```python
from pants.engine.rules import concurrently, rule

@rule
async def my_rule() -> MyResult:
    first_party_deps, third_party_deps = await concurrently(

>         FirstPartyDepsRequest(...), ThirdPartyDepsRequest(...)

Author: Whoops yes, the actual rule names are missing. Fixed. Good catch!
